configmap & utime sync: provide via hive cell

mhofstetter opened the pull request:
Currently, the interaction with the BPF map configmap and the corresponding utime sync is mainly via global accessors.

This commit refactors this, by introducing hive cells for the configmap and the utime sync.

mhofstetter committed:
configmap & utime sync: provide via hive cell

Currently, the interaction with the BPF map configmap and the
corresponding utime sync is mainly via global accessors.

This commit refactors this, by providing hive cells for the configmap
and the utime sync.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter commented on pkg/datapath/linux/utime/cell.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+func initUtimeSync(lifecycle hive.Lifecycle, configMap configmap.Map) {
+	controllerManager := controller.NewManager()
```
@joamaki would it be worth to have a provider for the controller manager?

joamaki replied:
What do you think of https://github.com/cilium/cilium/pull/24558?

mhofstetter replied:
interesting - will take a look.

i think something like this could help in this case - and could be refactored in the future (once its merged)

joamaki commented on pkg/maps/configmap/cell.go:
```diff
@@ -0,0 +1,36 @@
+	"eBPF map config contains runtime configuration state for the Cilium datapath",
+
+	cell.Provide(newMap),
+	cell.Invoke(func(Map) {}),
```
Forced instantiation really required?

mhofstetter replied:
no not really - having the utime sync depending on it - which triggers the instantiation.

but i thought it would be best to instantiate the map properly so the datapath can rely on it. otherwise it kind of always needs someone who depends on it to get initialized. or does datapath code create the map as well and would work without this initialization from userspace?

jrajahalme replied:
It looks to me that we do not create any bpf maps from the datapath?

mhofstetter replied:
introduced a bpf registry where map cells register themselves by using the value group `bpfMaps`.

this way datapath constructor can depend on the bpf map registry and the initialization order is ensured.

it's just a start which addresses this issue. might be enriched / replaced in the future.

@joamaki can you take another look :eyes:

joamaki commented on pkg/maps/configmap/cell.go:
```diff
@@ -0,0 +1,36 @@
+				return fmt.Errorf("failed to init config map: %w", err)
+			}
+
+			return nil
```
Stop hook missing that closes it

mhofstetter replied:
thanks - added!

jrajahalme reviewed (approved):
LGTM

mhofstetter committed:
datapath: depend on bpf maps managed by hive cell

This commit lets the datapath depend on all BPF maps which are managed
by a hive cell. BPF Map cells can register their map to the registry via
hive value group `bpf-maps` and the new BpfMap marker interface.

This way, we ensure that BPF maps are initialized via their cell
before the loader.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
authmap: register map

Currently, the initialization of the auth map is mainly driven by the
usage in the auth manager.

To always guarantee the initialization before the bpf loader, with this
commit, the auth map gets registered in the map registry.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
configmap: register map

To always guarantee the initialization of the config bpf map
before the bpf loader, with this commit, the auth map gets
registered in the map registry.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

joamaki commented on pkg/maps/mapregistry/cell.go:
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
```
I would put this file into `pkg/maps/cell.go`. I don't think we need a separate package for it. There it also makes more sense that it references all the maps underneath it.

joamaki replied:
To break the cycle to `BpfMap` move that to `pkg/maps/types` ?
EDIT: or to pkg/bpf. I think eventually that's where we'd have the canonical bpf map type anyway. e.g. I'd expect we'd have `Loader` depending on all the maps and then rewriting references using the open maps rather than pin paths, so this `BpfMap` becomes a real thing rather than a dependency marker.

mhofstetter replied:
yeah, the actual reason was the cycle. but makes much more sense by having cells in `pkg/maps` & the type in `pkg/bpf`. thanks

joamaki commented on pkg/maps/configmap/cell.go:
```diff
@@ -0,0 +1,43 @@
+	cell.Out
+
+	ConfigMap Map
+	BpfMap    maps.BpfMap `group:"bpfMaps"`
```
Convention for group struct tags is to use dashed names: `bpf-maps`

mhofstetter replied:
thanks, fixed

joamaki commented on pkg/maps/mapregistry/cell.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+// MapRegistry contains all BPF Maps which have been registered.
+type MapRegistry struct {
```
I would just add to datapathParams:
```
Maps []maps.BpfMap `group:"bpf-maps"`
```

This wrapper seems like an unnecessary indirection.

mhofstetter replied:
had it this way when starting out, but felt like exposing too many internals by depend on the value group name alone. but definitely agree that the indirection otherwise doesn't make sense -> removed

joamaki commented on pkg/maps/configmap/cell.go:
```diff
@@ -0,0 +1,43 @@
+	}
+}
+
+type MapOut struct {
```
I usually try to avoid writing `cell.Out`s with group values for the fear of typoing the group name. You can get around this with a helper, e.g. `func NewMapOut[Map any](m Map) MapOut[Map]`, but perhaps not worth adding it just yet.

lmb commented on pkg/maps/cells.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+// Cell contains all cells which are providing BPF Maps.
+var Cell = cell.Module(
```
When should I add something here vs. to the datapath cell as done for the configmap?

lmb replied:
Ah nvm utime.Cell is not a map.

mhofstetter replied:
exactly - utime just syncs the utime and uses the configmap to write from user- to kernelspace. 

the cell of the configmap itself is part of the maps cell.
